janusgraph-core/src/main/java/org/janusgraph/graphdb/util/MultiIterator.java

```diff
@@ -16,44 +16,46 @@
 
 import org.apache.tinkerpop.gremlin.structure.util.CloseableIterator;
 
+import java.util.Collection;
 import java.util.Iterator;
-import java.util.List;
-import java.util.Objects;
 
 public class MultiIterator<E> extends CloseableAbstractIterator<E> {
 
-    private Iterator<Iterator<E>> listIterator;
+    private final Iterator<Iterator<E>> listIterator;
     private Iterator<E> currentIterator;
 
-    public MultiIterator(final List<Iterator<E>> iterators) {
-        Objects.requireNonNull(iterators);
+    /**
+     * @param iterators iterators to concatenate. Must be not null. During iteration, iterators are closed lazily when
+     *                  exhausted (if an iterator extends {@link AutoCloseable}).
+     */
+    public MultiIterator(final Collection<Iterator<E>> iterators) {
         listIterator = iterators.iterator();
+        if(listIterator.hasNext()){
+            currentIterator = listIterator.next();
+        } else {
+            endOfData();
```

Comment:
I think this will lead to NPE when computeNext method is executed, because currentIterator is null.

Reply:
computeNext cannot be executed after `endOfData();` execution

Reply:
`computeNext` is a protected method to be executed by the abstract iterator only. After `endOfData()` - `computeNext()` won't be executed anymore

Reply:
Yeah, you are right, I missed that. The code looks good to me then. Can you add a test case for this? I checked and existing tests didn't cover this case.

Reply:
Btw it just came to my mind that the name `listIterator` does not quite make sense because the input is a Collection of iterators, not List of iterators.

Reply:
Added tests and renamed to `collectionIterator`